### PR TITLE
Nicer error messages, Selectable C/Z/T

### PR DESCRIPTION
**Pull request by manics**
Testing: Don't merge this yet, I'll setup ome-analysis-01 at some point.
- If you select an image without features you should get an error message instead of an exception.
- The feature calculation script allows C/Z/T to be chosen, defaults are all channels middle Z middle T.
- In the right hand pane clicking on the C.Z.T triplet for a reference image should make them editable. In practice you'll probably only want to change the channel unless you've specifically requested features for other Z/T.
- Multiple results for the same image (different C/Z/T) should be supported. This requires changing DOM IDs from object-id to object-id-extrastuff, and therefore requires PR https://github.com/openmicroscopy/openmicroscopy/pull/1234 to avoid breaking OMERO.web.
- The feature calculation script now has the option to not recalculate features if they are already present, default don't recalculate.

### CODE REVIEW

File: templates/searcher/plugin_config/right_search_form.js.html [hunk 1]
```diff
@@ -7,7 +7,16 @@
         load_tab_content: function(selected, obj_dtype, obj_id) {    // This MUST be defined
             var imageIds = [];
             for (var i=0; i<selected.length; i++) {
-                imageIds.push(selected[i]["id"].replace("-","="));
+		// IDs will be in the form image-<id> if coming from the
+		// standard middle pane, or image-<id>-<superid> if coming
+		// from the OMERO.searcher middle pane
+		sid = selected[i]["id"].split('-');
```
Comment (will-moore) [inline templates/searcher/plugin_config/right_search_form.js.html:13]: should that be:
var sid = selected[i]["id"].split('-');

**Comment by pwalczysko** [comment]
`If you select an image without features you should get an error message instead of an exception.` - this works, just the font size of the error message is too small
screen shot 2013-06-10 at 15 05 10

**Comment by pwalczysko** [comment]
Re: `The feature calculation script allows C/Z/T to be chosen, defaults are all channels middle Z middle T.`
BUG: Stacktrace after feature calculation script is run with defaults.
 Workflow: 
- imported images from test_images_good/avi.
- select the first .avi
- Go Scripts > Searcher > Omero Searcher Feature Calculation...
- In the dialog (see below), leave everything at default (although there are `-1` values for Z and T (see below)
- Click `Run script`
- the Activities reports a green tick, but in the `!` there is a stacktrace  to be found

screen shot 2013-06-10 at 17 15 08

```
/usr/lib64/python2.6/site-packages/mahotas/texture.py:6: DeprecationWarning: Use

from mahotas.features import texture

  ''', DeprecationWarning)
/usr/lib64/python2.6/site-packages/mahotas/zernike.py:6: DeprecationWarning: Use

from mahotas.features import zernike

  ''', DeprecationWarning)
/usr/lib64/python2.6/site-packages/mahotas/tas.py:6: DeprecationWarning: Use

from mahotas.features import tas

  ''', DeprecationWarning)
/usr/lib64/python2.6/site-packages/mahotas/lbp.py:6: DeprecationWarning: Use

from mahotas.features import lbp

  ''', DeprecationWarning)
WARNING:omero.gateway:ApiUsageException on <class 'omero.gateway.OmeroGatewaySafeCallWrapper'> to <0ecbd687-b5f0-4c42-88e2-278700a13015omero.api.RawPixelsStore> getPlane((111, 0, 0), {})
Traceback (most recent call last):
  File "/opt/omero/server/lib/python/omero/gateway/__init__.py", line 3457, in __call__
    return self.f(*args, **kwargs)
  File "/opt/omero/server/lib/python/omero_api_RawPixelsStore_ice.py", line 1069, in getPlane
    return _M_omero.api.RawPixelsStore._op_getPlane.invoke(self, ((z, c, t), _ctx))
ApiUsageException: exception ::omero::ApiUsageException
{
    serverStackTrace = ome.conditions.ApiUsageException: Z '111' greater than sizeZ '1'.
    at ome.services.RawPixelsBean.handleException(RawPixelsBean.java:681)
    at ome.services.RawPixelsBean.getPlane(RawPixelsBean.java:388)
    at sun.reflect.GeneratedMethodAccessor709.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:601)
    at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
    at ome.security.basic.EventHandler.invoke(EventHandler.java:154)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
    at ome.tools.hibernate.SessionHandler.doStateful(SessionHandler.java:182)
    at ome.tools.hibernate.SessionHandler.invoke(SessionHandler.java:166)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
    at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
    at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:241)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
    at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:116)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
    at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
    at com.sun.proxy.$Proxy102.getPlane(Unknown Source)
    at sun.reflect.GeneratedMethodAccessor709.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:601)
    at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
    at ome.security.basic.BasicSecurityWiring.invoke(BasicSecurityWiring.java:98)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
    at ome.services.blitz.fire.AopContextInitializer.invoke(AopContextInitializer.java:43)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
    at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
    at com.sun.proxy.$Proxy102.getPlane(Unknown Source)
    at sun.reflect.GeneratedMethodAccessor795.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:601)
    at ome.services.blitz.util.IceMethodInvoker.invoke(IceMethodInvoker.java:179)
    at ome.services.throttling.Callback.run(Callback.java:56)
    at ome.services.throttling.InThreadThrottlingStrategy.callInvokerOnRawArgs(InThreadThrottlingStrategy.java:56)
    at ome.services.blitz.impl.AbstractAmdServant.callInvokerOnRawArgs(AbstractAmdServant.java:150)
    at ome.services.blitz.impl.RawPixelsStoreI.getPlane_async(RawPixelsStoreI.java:101)
    at sun.reflect.GeneratedMethodAccessor794.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:601)
    at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
    at omero.cmd.CallContext.invoke(CallContext.java:59)
    at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
    at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
    at com.sun.proxy.$Proxy103.getPlane_async(Unknown Source)
    at omero.api._RawPixelsStoreTie.getPlane_async(_RawPixelsStoreTie.java:143)
    at omero.api._RawPixelsStoreDisp.___getPlane(_RawPixelsStoreDisp.java:1041)
    at omero.api._RawPixelsStoreDisp.__dispatch(_RawPixelsStoreDisp.java:1468)
    at IceInternal.Incoming.invoke(Incoming.java:159)
    at Ice.ConnectionI.invokeAll(ConnectionI.java:2357)
    at Ice.ConnectionI.dispatch(ConnectionI.java:1208)
    at Ice.ConnectionI.message(ConnectionI.java:1163)
    at IceInternal.ThreadPool.run(ThreadPool.java:302)
    at IceInternal.ThreadPool.access$300(ThreadPool.java:12)
    at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:643)
    at java.lang.Thread.run(Thread.java:722)

    serverExceptionClass = ome.conditions.ApiUsageException
    message = Z '111' greater than sizeZ '1'.
}
Traceback (most recent call last):
  File "./script", line 329, in <module>
    runScript()
  File "./script", line 317, in runScript
    message += processImages(client, scriptParams) + '\n'
  File "./script", line 189, in processImages
    channels, zselect, tselect, recalc)
  File "./script", line 146, in extractFeatures
    conn, image, scale, ftset, scaleSet, chs, zslice, timepoint)
  File "./script", line 49, in extractFeaturesOneChannel
    pixels, channels, zslice, timepoint, debug=True)
  File "/usr/lib/python2.6/site-packages/pyslid/features.py", line 245, in calculate
    img.channeldata[ 'protein' ] = scipy.misc.imresize( plane, scale )
  File "/usr/lib64/python2.6/site-packages/scipy/misc/pilutil.py", line 224, in imresize
    im = toimage(arr)
  File "/usr/lib64/python2.6/site-packages/scipy/misc/pilutil.py", line 94, in toimage
    assert valid, "Not a suitable array shape for any mode."
AssertionError: Not a suitable array shape for any mode.
```

**Comment by pwalczysko** [comment]
Re: `In the right hand pane clicking on the C.Z.T triplet for a reference image should make them editable. In practice you'll probably only want to change the channel unless you've specifically requested features for other Z/T`
- this works (see below) but
- I cannot execute the Search as
- I do not have any features calculated for a multi Z T image
- I cannot calculate any new features because of the bug in the previous comment

screen shot 2013-06-10 at 17 22 14

**Comment by pwalczysko** [comment]
Re: `Multiple results for the same image (different C/Z/T) should be supported. This requires changing DOM IDs from object-id to object-id-extrastuff, and therefore requires PR openmicroscopy/openmicroscopy#1234 to avoid breaking OMERO.web.` - not sure how to test this, see the blocking bug in one but last comment

**Comment by pwalczysko** [comment]
Re: `The feature calculation script now has the option to not recalculate features if they are already present, default don't recalculate.`
Yes, there is a box which you can tick. But after the `Run script` is clicked, again the bug from above appears.

**Comment by manics** [comment]
Thanks @pwalczysko. The deprecation warnings are due to one of the dependent Python modules from CMU (Pyslid/Mahotas), I've added [Trac #11099](https://trac.openmicroscopy.org.uk/ome/ticket/11099). The warnings shouldn't affect feature calculation, though they will always result in output to STDERR which means the `!` will always show.

I've investigated the stacktrace, which is due to a bug in Pyslid (C/Z/T indicies are mixed up): [Trac #11098](https://trac.openmicroscopy.org.uk/ome/ticket/11098)

**Pull request merged**